smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/validators/EndpointModifierValidator.java

```diff
@@ -0,0 +1,100 @@
+ * Validate EndpointModifier traits that are applied on a service.
+ */
+public class EndpointModifierValidator extends AbstractValidator {
+
+    private static final Set<String> SUPPORTED_PATTERNS = SetUtils.of(
+            "{region}", "{service}", "{dnsSuffix}", "{dualStackDnsSuffix}"
+    );
+
+    private static final Pattern PATTERN = Pattern.compile("\\{[^\\}]*\\}");
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        for (ServiceShape serviceShape : model.getServiceShapesWithTrait(StandardRegionalEndpointsTrait.class)) {
+            events.addAll(validateRegionalEndpointPatterns(serviceShape,
+                    serviceShape.expectTrait(StandardRegionalEndpointsTrait.class)));
+        }
+        for (ServiceShape serviceShape : model.getServiceShapesWithTrait(StandardPartitionalEndpointsTrait.class)) {
+            events.addAll(validatePartitionalEndpointPatterns(serviceShape,
+                    serviceShape.expectTrait(StandardPartitionalEndpointsTrait.class)));
+        }
+        return events;
+    }
+
+    private List<ValidationEvent> validateRegionalEndpointPatterns(
+            ServiceShape serviceShape, StandardRegionalEndpointsTrait regionalEndpoints) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        regionalEndpoints.getRegionSpecialCases().forEach((region, specialCases) -> {
+            specialCases.forEach((specialCase) -> {
+                events.addAll(validateEndpointPatterns(
+                        serviceShape, regionalEndpoints, specialCase.getEndpoint()));
+            });
+        });
+
+        regionalEndpoints.getPartitionSpecialCases().forEach((region, specialCases) -> {
+            specialCases.forEach((specialCase) -> {
+                events.addAll(validateEndpointPatterns(
+                        serviceShape, regionalEndpoints, specialCase.getEndpoint()));
+            });
+        });
+
+        return events;
+    }
+
+    private List<ValidationEvent> validatePartitionalEndpointPatterns(
+            ServiceShape serviceShape, StandardPartitionalEndpointsTrait partitionalEndpoints) {
+
+        List<ValidationEvent> events = new ArrayList<>();
+
+        partitionalEndpoints.getPartitionEndpointSpecialCases().forEach((region, specialCases) -> {
+            specialCases.forEach((specialCase) -> {
+                events.addAll(validateEndpointPatterns(
+                        serviceShape, partitionalEndpoints, specialCase.getEndpoint()));
+            });
+        });
+
+        return events;
+    }
+
+    private List<ValidationEvent> validateEndpointPatterns(
+            ServiceShape serviceShape, FromSourceLocation location, String endpoint) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        Matcher m = PATTERN.matcher(endpoint);
+        while (m.find()) {
+            if (!SUPPORTED_PATTERNS.contains(m.group())) {
+                events.add(danger(
```

Comment:
Can these be aggregated to a list so that one event is emitted for an invalid endpoint pattern with any amount of unsupported labels in it?

Reply:
Yeah - will combine.